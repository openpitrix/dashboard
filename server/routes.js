const Router = require('koa-router');
const request = require('lib/request').default;
const { getServerConfig } = require('lib/utils');

const router = new Router();
const config = getServerConfig();

router.get('/api/:version/*', async ctx => {
  try {
    ctx.body = await request.get(config.serverUrl + ctx.url, ctx.query);
  } catch (err) {
    ctx.status = err.code || 404;
  }
});

router.post('/login', async ctx => {
  const resp = await request.post(`${config.serverUrl}/api/v1/login`, ctx.request.body);
  if (resp.ret_code === 0) {
    // todo: handle session
    ctx.redirect('/');
  } else {
    ctx.body = resp;
  }
});

// fallback to handle all request
// router.all('*', (ctx, next)=> {
//   console.log('ctx.url', ctx.url);
//   console.log('ctx.query', ctx.query);
//
//   ctx.body = 'Bad request';
//   ctx.status = 403;
// });

module.exports = router;
