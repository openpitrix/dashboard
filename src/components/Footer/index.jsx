import React, { PureComponent } from 'react';
import { translate } from 'react-i18next';

import styles from './index.scss';

@translate()
export default class Footer extends PureComponent {
  getCurrentLocale = () => typeof window !== 'undefined' && localStorage.getItem('i18nextLng');

  changeLocale = (lang, e) => {
    e.preventDefault();
    if (this.getCurrentLocale() !== lang) {
      this.props.i18n.changeLanguage(lang);
    }
  };

  render() {
    const { t } = this.props;

    return (
      <div className={styles.footer}>
        <div className={styles.wrapper}>
          <div className={styles.inner}>
            <span className={styles.logo}>
              <img src="/logo_grey.svg" alt="logo" height="100%" />
            </span>
            <ul className={styles.terms}>
              <li>
                <a
                  target="_blank"
                  href="https://docs.openpitrix.io/home/"
                  rel="noopener noreferrer"
                >
                  {t('Document')}
                </a>
              </li>
              <li>
                <a
                  target="_blank"
                  href="https://openpitrix.io"
                  rel="noopener noreferrer"
                >
                  {t('About')}
                </a>
                {/* <span className={styles.dot} />
                 <a href="javascript:void(0)">{t('Help')}</a>
                 <span className={styles.dot} />
                 <a href="javascript:void(0)">{t('Terms')}</a> */}
              </li>
              <li>OpenPitrix &copy; 2018</li>
              <li>
                <a
                  href="#"
                  onClick={this.changeLocale.bind(null, 'zh')}
                  className={styles.toggleLocale}
                >
                  中文
                </a>
                <span className={styles.dot} />
                <a
                  href="#"
                  onClick={this.changeLocale.bind(null, 'en')}
                  className={styles.toggleLocale}
                >
                  English
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    );
  }
}
