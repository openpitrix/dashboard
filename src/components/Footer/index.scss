@import "~scss/vars";

$line-height: $footer-height - 1;
$fg-color: #78839e;

.footer {
  position: static;
  width: 100%;
  height: $footer-height;
  background-color: $N0;
  box-sizing: border-box;

  .wrapper {
    width: $content-width;
    margin: 0 auto;
    color: $fg-color;

    .logo {
      display: inline-block;
      float: left;
      font-size: $size-medium;
      line-height: $line-height;
      text-transform: uppercase;
      :global {
        img {
          height: 16px;
        }
      }
    }
    .terms {
      display:inline-block;
      float: right;
      line-height: $line-height;
      & > li{
        list-style-type: none;
        float: left;
        text-align: right;
        margin-left: 10px;

        &.copyright {
          margin-left: 20px;
        }

        a {
          color: $fg-color;
          &:hover {
            color: $P75;
          }
        }
      }
    }
  }

  .dot {
    display:inline-block;
    height: 3px;
    width: 3px;
    background-color: $fg-color;
    border-radius: 50%;
    vertical-align: middle;
  }

  .toggleLocale {
    &:hover {
      color: $P75 !important;
    }
  }
}
