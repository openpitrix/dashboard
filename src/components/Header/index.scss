@import "~scss/vars";

$profile-hover-color: #efe6f8;

@keyframes deepHeader {
  from {
    opacity: 0;
    background: transparent;
  }
  to {
    opacity: 1;
    background: linear-gradient(to left, $P75, #854fb9 32%, #484999);
  }
}

.header {
  position: absolute;
  top: 0;
  width: 100%;
  height: $header-height;
  z-index: 999;
  background: none;

  &.deep {
    background: linear-gradient(to left, $P75, #854fb9 32%, #484999);
  }

  &.deep, &.deepHome {
    position: fixed;

    .login{
      color: $N0;
    }

    .user [class*='target']{
      color: $N0;

      :global{
        .icon-caret-down svg{
          --primary-color: #{$N0};
          --secondary-color: #{$N0};
        }
      }
    }
  }


  &.deepHome {
    animation-duration: 0.5s;
    animation-name: deepHeader;
    animation-iteration-count: 1;
    animation-timing-function: ease-in-out;
    animation-fill-mode: forwards;
  }

  .wrapper {
    width: $content-width;
    height: $header-height;
    margin: 0 auto;
  }

  .logo {
    float: left;
    width: auto;
    height: $logo-height;
    display: inline-block;
    margin-top: $header-height / 2;
    transform: translateY(-50%);
    vertical-align: middle;
  }

  .search {
    float: right;
    display: inline-block;
    min-width: 240px;
    margin-right: 20px;
    margin-top: $header-height / 2;
    transform: translateY(-50%);
    vertical-align: top;

    input {
      box-sizing: border-box;
      width: 240px;
      color: $P45;
      font-weight: 400;
      font-size: $size-normal;
      border-radius: 16px;
      border: 1px solid transparent;
      background-color: rgba($N0, 0.1);
      //transition: color .5s ease-in;
      transition: all .3s ease-in-out;

      &::placeholder {
        font-size: 14px;
        color: $P45;
        line-height: 1.0;
        letter-spacing: 0;
      }

      &:focus {
        transition: width .2s ease-in;
        width: 360px;
        font-weight: normal;
        color: $N0;
        border: solid 1px rgba($N0, 0.5) !important;
      }
      &:hover {
        color: $N0;
        border: solid 1px rgba($N0, 0.3);
      }
    }

    :global {
      .qicon {
        --primary-color: #fff;
        //--secondary-color: #d3c4ed;
      }
      .icon {
        color: $N0;
        opacity: 0.5;
        &:hover, &:focus {
          opacity: 1;
          svg {
            --primary-color: #fff;
          }
        }
      }
    }
  }

  .login{
    float: right;
    display: inline-block;
    margin: 14px 0;
    line-height: 20px;
    color: $N500;
    font-weight: 500;
    opacity: 0.9;

    &:hover{
      opacity: 1;
    }
  }

  .user [class*='target']{
    float: right;
    margin: 12px 24px;
    height: 20px;
    width: auto;
    line-height: 20px;
    opacity: 0.9;
    color: $N500;
    font-weight: 500;
    border: 0 none;

    &:hover{
      border: 0 none;
      box-shadow: none;
      opacity: 1;
    }

    :global{
      .icon-caret-down{
        position: relative;
        top: 3px;
        left: 2px;
        svg {
          --primary-color: #{$N500};
          --secondary-color: #{$N500};
        }
      }
    }
  }

  .operateItems{
    min-width: 120px;
    li a{
      display: block;
      height: 32px;
      padding: 0 12px;
      //width: 100%;
      line-height: 32px;
      font-size: $font14;
      font-weight: normal;
      color: $N300;
      border-radius: 0;
      &:hover {
        background-color: $N10;
        color: $N500;
      }
      &.line{
        border-bottom: 1px solid $N10;
      }
    }
  }

  :global{
    .pi-popover-popper{
      margin-left: -62px !important;
      margin-top: 6px !important;
    }
  }
}

.menuOuter {
  .menus{
    float: left;
    margin-left: 50px;
    a{
      display: inline-block;
      margin: 0 12px;
      padding: 13px 0;
      font-size: 14px;
      font-weight: 500;
      line-height: 20px;
      color: $N0;
      &.active{
        border-bottom: 2px solid $Y200;
      }
    }
  }
}
