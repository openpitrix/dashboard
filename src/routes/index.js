import { lazy } from 'react';

// views without lazy load
import Home from 'pages/Home';

import * as Dash from 'pages/Dashboard';

// views using lazy load
const Login = lazy(() => import('../pages/Login'));
const AppDetail = lazy(() => import('../pages/AppDetail'));
const Profile = lazy(() => import('../pages/Profile'));
const SSHKeys = lazy(() => import('../pages/SSHKeys'));
const Store = lazy(() => import('../pages/Store'));
const Purchased = lazy(() => import('../pages/Purchased'));
const Runtimes = lazy(() => import('../pages/Runtimes'));

const dashboardPrefix = '/dashboard';

const dashUrl = path => {
  if (!path) return dashboardPrefix;
  if (path.startsWith('/:dash')) {
    path = path.replace('/:dash', dashboardPrefix);
  }
  return path;
};

const routes = {
  '/': Home,
  '/login': Login,
  '/apps/search/:search': Home,
  '/apps/category/:category': Home,
  '/apps/:appId': AppDetail,

  '/store': Store,
  '/store/search/:search': Store,
  '/store/category/:category': Store,
  '/store/:appId': AppDetail,
  '/store/:appId/deploy': Dash.AppDeploy,

  '/purchased': Purchased,
  '/purchased/:clusterId': Dash.ClusterDetail,

  '/runtimes': Runtimes,

  '/:dash': Dash.Overview,

  '/:dash/apps': Dash.Apps,
  '/dev/apps': Dash.Apps,
  '/:dash/reviews': Dash.AppReview,
  '/:dash/review/:appId/:versionId': AppDetail,
  '/:dash/app/create': Dash.AppAdd,
  '/:dash/app/:appId': Dash.AppDetail,
  '/:dash/audit/record/:appId': Dash.AuditRecord,

  '/:dash/clusters': Dash.Clusters,

  '/:dash/cluster/:clusterId': Dash.ClusterDetail,

  '/:dash/runtimes': Dash.Runtimes,
  '/:dash/runtime/create': Dash.RuntimeAdd,
  '/:dash/runtime/edit/:runtimeId': Dash.RuntimeAdd,
  '/:dash/runtime/:runtimeId': Dash.RuntimeDetail,

  '/:dash/repos': Dash.Repos,
  '/:dash/repo/create': Dash.RepoAdd,
  '/:dash/repo/edit/:repoId': Dash.RepoAdd,
  '/:dash/repo/:repoId': Dash.RepoDetail,

  '/:dash/categories': Dash.Categories,
  '/:dash/category/:categoryId': Dash.CategoryDetail,

  '/:dash/users': Dash.Users,
  '/:dash/user/:userId': Dash.UserDetail,

  '/profile': Profile,
  '/ssh_keys': SSHKeys,

  '*': Home
};

const judgeNeedAuth = route => {
  if (route.startsWith('/:dash')) {
    return true;
  }

  if (route.startsWith('/store/:appId/deploy')) {
    return true;
  }

  const paths = ['/runtimes', '/purchased', '/profile', '/ssh_keys', '/store'];
  return paths.includes(route);
};

export default Object.keys(routes).map(route => {
  const routeDefinition = Object.assign(
    {},
    {
      path: dashUrl(route),
      exact: route !== '/login',
      component: routes[route],
      needAuth: judgeNeedAuth(route)
    }
  );
  if (route === '*') {
    Object.assign(routeDefinition, { noMatch: true });
  }
  return routeDefinition;
});
