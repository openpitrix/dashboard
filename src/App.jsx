import React, { Fragment, Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router';
import { Switch, Route, Redirect } from 'react-router-dom';
import { Provider } from 'mobx-react';

import Header from 'components/Header';
import Footer from 'components/Footer';
import RouteWrapper from './routes/wrapper';
import routes from './routes';
import { getCookie } from './utils';

import './scss/main.scss';

class App extends Component {
  static propTypes = {
    rootStore: PropTypes.object
  };

  hasLoggedIn = () => {
    return !!getCookie('user');
  };

  render() {
    const { location, rootStore } = this.props;
    const pathname = location.pathname;

    const isHome = pathname === '/' || pathname === '/apps';
    const isLogin = pathname === '/login';

    return (
      <Provider rootStore={rootStore}>
        <Fragment>
          {!isLogin && <Header isHome={isHome} />}
          <div className="main">
            <Switch>
              {routes.map((route, i) => (
                <Route
                  key={i}
                  exact={route.exact}
                  path={route.path}
                  render={({ match }) =>
                    route.needAuth && !this.hasLoggedIn() ? (
                      <Redirect to="/login" />
                    ) : (
                      <RouteWrapper
                        component={route.component}
                        match={match}
                        rootStore={rootStore}
                      />
                    )
                  }
                />
              ))}
            </Switch>
          </div>
          {!isLogin && <Footer />}
        </Fragment>
      </Provider>
    );
  }
}

export default withRouter(App);
