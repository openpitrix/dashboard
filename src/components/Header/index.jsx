import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { NavLink } from 'react-router-dom';
import { observer, inject } from 'mobx-react';
import { ButtonDropdown, DropdownToggle, DropdownItem, DropdownMenu } from 'reactstrap';
import { getLoginUser } from 'src/utils';

import Logo from '../Logo';
import Input from '../Base/Input';
import styles from './index.scss';

@inject('rootStore')
@observer
export default class Header extends Component {
  state = {
    isHome: PropTypes.bool,
    dropdownOpen: false
  };

  toggleDropdown = () => {
    this.setState({
      dropdownOpen: !this.state.dropdownOpen
    });
  };

  renderLoginButton() {
    const loggeInUser = getLoginUser();

    if (!loggeInUser) {
      return <NavLink to="/login">Sign In</NavLink>;
    }

    return (
      <ButtonDropdown
        className={styles.loginButton}
        isOpen={this.state.dropdownOpen}
        toggle={this.toggleDropdown}
      >
        <DropdownToggle caret className={styles.toggleBtn}>
          {loggeInUser}
        </DropdownToggle>
        <DropdownMenu>
          <DropdownItem style={{ padding: 0 }}>
            <a href="/logout">Logout</a>
          </DropdownItem>
        </DropdownMenu>
      </ButtonDropdown>
    );
  }

  render() {
    const {
      isHome,
      rootStore: { fixNav }
    } = this.props;
    const isDark = !isHome || fixNav;
    const logoUrl = isDark ? '/assets/logo2.svg' : '/assets/logo.svg';

    return (
      <div className={classnames('header', styles.header, { [styles.darkHeader]: !isDark })}>
        <div className={styles.wrapper}>
          <Logo className={styles.logo} url={logoUrl} />
          {isDark && (
            <Input.Search className={styles.search} placeholder="Search apps in Pitrix..." />
          )}
          <div className={styles.menus}>
            <NavLink to="/apps">Catalog</NavLink>
            <NavLink to="/deployment">Deployment</NavLink>
            <NavLink to="/manage">{getLoginUser() ? 'Manage' : 'Development'}</NavLink>
            {this.renderLoginButton()}
          </div>
        </div>
      </div>
    );
  }
}
