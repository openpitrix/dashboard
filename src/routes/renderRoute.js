import React, { Fragment } from 'react';
import { Redirect } from 'react-router-dom';
import { withRouter } from 'react-router';

import Header from 'components/Header';
import Footer from 'components/Footer';
import RouteWrapper from './wrapper';
import { getCookie, setCookie } from '../utils';

const renderRoute = (match, route, store) => {
  if (route.path === 'login') {
    const names = [
      'access_token',
      'token_type',
      'access_token_home',
      'refresh_token',
      'user',
      'role'
    ];
    names.forEach(name => setCookie(name, '', -1));
  }

  const user = store.user || {};
  const role = getCookie('role');
  const hasHeader = user.isNormal || !user.username || role === 'user';

  if (route.needAuth && !user.username) {
    return <Redirect to={`/login?url=${match.url}`} />;
  }

  if (route.noMatch) {
    return <Redirect to="/" />;
  }

  // attach history to component
  const component = withRouter(route.component);

  if (route.path !== '/login') {
    const isHome = route.path === '/' || route.path.startsWith('/apps/');

    return (
      <Fragment>
        {(hasHeader || isHome) && <Header isHome={isHome} />}
        <RouteWrapper component={component} match={match} rootStore={store} />
        {(hasHeader || isHome) && <Footer />}
      </Fragment>
    );
  }
  return <RouteWrapper component={component} match={match} rootStore={store} />;
};

export default renderRoute;
