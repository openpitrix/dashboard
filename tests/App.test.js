import { observable, useStrict } from 'mobx';
import { Provider } from 'mobx-react';
// import { withRouter } from 'react-router';
import { BrowserRouter, Switch, Route, Redirect } from 'react-router-dom';

import App from 'src/App';
import routes from 'src/routes';
import renderRoute from 'src/routes/renderRoute';
import Header from 'components/Header';
import Footer from 'components/Footer';

describe('<App/>', () => {
  let store, location;

  beforeEach(() => {
    useStrict(false);

    store = observable({
      fixNav: false
    });
  });

  afterEach(() => {
    store = null;
  });

  const renderPage = (location, store) => {
    return (
      <Provider rootStore={store}>
        <BrowserRouter>
          <App location={location}>
            <Switch>
              {routes.map((route, i) => (
                <Route
                  key={i}
                  exact={route.exact}
                  path={route.path}
                  render={({ match }) => renderRoute(match, route, store)}
                />
              ))}
            </Switch>
          </App>
        </BrowserRouter>
      </Provider>
    );
  };

  it('basic render', () => {
    location = { pathname: '/' };
    const wrapper = render(renderPage(location, store));
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('body has main className', () => {
    const wrapper = shallow(<App rootStore={store} />);
    expect(wrapper.find('.main').length).toBe(1);
  });

  it('/login page does not have Header and Footer', () => {
    location = { pathname: '/login' };
    const wrapper = shallow(<App rootStore={store} location={location} />);
    expect(wrapper.find(Header).length).toBe(0);
    expect(wrapper.find(Footer).length).toBe(0);
  });

  it('page except /login have both Header and Footer', () => {
    location = { pathname: '/' };
    const wrapper = shallow(<App rootStore={store} location={location} />);
    expect(wrapper.find(Header).length).toBe(1);
    expect(wrapper.find(Footer).length).toBe(1);
  });
});
