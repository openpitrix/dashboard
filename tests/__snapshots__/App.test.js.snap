// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> basic render 1`] = `
Array [
  <div
    class="header header darkHeader"
  >
    <div
      class="wrapper"
    >
      <a
        aria-current="false"
        class="logo"
        href="/"
      >
        <img
          alt="logo"
          height="100%"
          src="/assets/logo_dark.svg"
        />
      </a>
      <div
        class="menus"
      >
        <a
          aria-current="false"
          href="/apps"
        />
        <a
          aria-current="false"
          href="/dashboard"
        />
        <a
          aria-current="false"
          href="/login"
        />
        <div
          class="translation dropdown"
        >
          <span
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle"
            data-toggle="dropdown"
            style="cursor:pointer"
          >
            <span
              class="sr-only"
            >
              Toggle Dropdown
            </span>
          </span>
          <div
            aria-hidden="true"
            class="transMenu dropdown-menu"
            role="menu"
            tabindex="-1"
          >
            <div
              class="locale"
            >
              <a
                href="#"
              />
            </div>
            <div
              class="locale"
            >
              <a
                href="#"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    class="main"
  />,
  <div
    class="footer"
  >
    <div
      class="wrapper"
    >
      <span
        class="logo"
      >
        <img
          alt="logo"
          height="100%"
          src="/assets/logo_grey.svg"
        />
      </span>
      <ul
        class="terms"
      >
        <li>
          <a
            href="javascript:;"
          >
            About
          </a>
        </li>
        <li>
          <a
            href="javascript:;"
          >
            Help
          </a>
        </li>
        <li>
          <a
            href="javascript:;"
          >
            Terms
          </a>
        </li>
        <li
          class="copyright"
        >
          OpenPitrix Â© 2018
        </li>
      </ul>
    </div>
  </div>,
]
`;
